pycytominer_transform/convert.py

```diff
@@ -0,0 +1,582 @@
+        # infer common basis of schema and column order for all others
+        for schema in [
+            parquet.read_schema(record["destination_path"]) for record in record_group
+        ]:
+
+            # account for completely equal schema
+            if schema.equals(writer_basis_schema):
+                continue
+
+            # gather field names from schema
+            schema_field_names = [item.name for item in schema]
+
+            # reversed enumeration because removing indexes ascendingly changes schema field order
+            for index, field in reversed(list(enumerate(writer_basis_schema))):
+
+                # check whether field name is contained within writer basis, remove if not
+                # note: because this only checks for naming, we defer to initially detected type
+                if field.name not in schema_field_names:
+
+                    writer_basis_schema = writer_basis_schema.remove(index)
+
+                # check if we have an integer to float challenge and enable later casting
+                elif pa.types.is_integer(field.type) and pa.types.is_floating(
+                    schema.field(field.name).type
+                ):
+                    writer_basis_schema = writer_basis_schema.set(
+                        index, field.with_type(pa.float64())
+                    )
+
+    if len(list(writer_basis_schema.names)) == 0:
+        raise Exception(
+            (
+                "No common schema basis to perform concatenation for record group."
+                " All columns mismatch one another within the group."
+            )
+        )
+
+    # build a parquet file writer which will be used to append files
+    # as a single concatted parquet file, referencing the first file's schema
+    # (all must be the same schema)
+    writer = parquet.ParquetWriter(str(destination_path), writer_basis_schema)
+
+    for table in [record["destination_path"] for record in record_group]:
+
+        # if we haven't inferred the common schema
+        # check that our file matches the expected schema, otherwise raise an error
+        if not infer_common_schema and not writer_basis_schema.equals(
+            parquet.read_schema(table)
+        ):
+            raise Exception(
+                (
+                    f"Detected mismatching schema for target concatenation group members:"
+                    f" {str(record_group[0]['destination_path'])} and {str(table)}"
+                )
+            )
+
+        # read the file from the list and write to the concatted parquet file
+        # note: we pass column order based on the first chunk file to help ensure schema
+        # compatibility for the writer
+        writer.write_table(parquet.read_table(table, schema=writer_basis_schema))
+        # remove the file which was written in the concatted parquet file (we no longer need it)
+        pathlib.Path(table).unlink()
+
+    # close the single concatted parquet file writer
+    writer.close()
```

Comment:
I'd highly recommend using a context manager here, e.g. `with parquet.ParquetWriter(str(destination_path), writer_basis_schema) as writer: <etc.>`. There are various ways that this code could exit abnormally, including an exception that you're raising yourself, which could lead to file corruption if `writer.close()` is never reached.

Reply:
Great point, thank you - I'll update this with a context manager instead.